### ISSUE TITLE
dynamic filter number range (filter condition `range`)(is_within)

### ISSUE THREAD

**Issue by pjmuller** [avo-hq/avo#3013]
### Feature
For dynamic_filter of type `date` we have the nice `is_within`, however for `number` we would also like the same `is_within` operator giving us 2 separate numeric inputs

There currently is no workaround because if you add 2x the same field, you get (rightfully for other types) the `OR` operator. For numbers this results to `my_number > 5` OR `my_number < 10` which returns ALL records as we needed an `AND`

So in short, the `OR` is the right operator for almost all field types, by adding the `is_within` filter we only need to register 1 filter which solves all problems :)

**Comment by Paul-Bob** [comment]
Makes total sense, we have the [group filter feature](https://github.com/avo-hq/avo/issues/2981) in mind that would also fix this issue but `is_within` applies in totality to number filters.

**Comment by adrianthedev** [comment]
I start to see a pattern.
Maybe the conditions should not be an array of labels or `{key => label}` like we'd use on the `select` input, but have more advanced objects with settings and stuff.

```ruby
options: [{
  id: :is_within,
  label: "Is within",
  is_range: true,
  small_label: "> x <" # this is displayed on the pill when the filter is collapsed
}, {
  id: is_greater,
  label: "Greater than",
  small_label: ">="
}]
```

But maybe still have it work with a simple array or select-styled options.

```ruby
[:is_within, :is_greater]

# or

{is_within: "Is within", is_greater: "Greater than"}
```

WDYT?

**Comment by Paul-Bob** [comment]
> WDYT?

`is_range: true` needs a default range input for date, other for time and other for number

How would this work with the fields DSL?:

```ruby
dynamic_filter :name,
  ...,
  fields: -> { # This gets overriden when a condition with is_range: true is selected?

  }
```

IMHO custom dynamic filter configurations are getting bigger and bigger and should be moved to a separate file.

A custom filter can easily reach 30+ lines with this DSL

This should be more organized:

```ruby

class Avo::Resources::User < Avo::BaseResource
  # ...
  def filters
    custom_dynamic_filter Avo::DynamicFilters::AgeBetween
  end
  # ...
end

class Avo::DynamicFilters::AgeBetween < Avo::DynamicFilters::Base
  # Figure out how to configure all options here
  # Maybe:
  def conditions
    [
      {
        id: :is_within,
        label: "Is within",
        is_range: true,
        small_label: "> x <" # this is displayed on the pill when the filter is collapsed
      },
      {
        id: is_greater,
        label: "Greater than",
        small_label: ">="
      }
    ]
  end
end
```

**Comment by adrianthedev** [comment]
Yes, we could do that, but let's first define what we need.

Regardign the API for the fields, we don't expose the `fields` option but take it upon us to generate two inputs.
If the filter is text, we would normally generate one text input.
If the filter is text and the condition is a range, we will generate two text inputs and those would give two params instead of one. See the API proposal.

**Comment by jetienne** [comment]
@adrianthedev Commenting here, following our discussion on Discord. I am looking to implement a filtering on a non existing field (My model has a method Villa.available_between(d1, d2), and a complex query in it). Ideally, I would like the users to select the field, then use the is_within operator and select a range. Is it already possible ? Any workaround ?

**Comment by adrianthedev** [comment]
Not available right now but I'm going to paste the code that should make it possible when we add the feature.

My suggestion is to make `available_between` a scope so it can be chained to the query.

```ruby
dynamic_filter :some_name,
  type: :number, # or :text
  conditions: [{
    id: :available_between,
    label: "Available between",
    is_range: true,
  }],
  query: -> {
    query.available_between(params[:available_between][:start], params[:available_between][:end])
  }
```

Now, writing this I have a few thoughts, maybe we should not render the conditions dropdown if there's only one(easy way). Or maybe have a `visible` option that takes a block (more complex).

The `is_range: true` should render two fields instead of one and pass those in the params.

**Comment by jetienne** [comment]
@adrianthedev Thank you. Do you have a rough non-binding ETA in mind?

**Comment by adrianthedev** [comment]
Very non-binding, late-August, early-September the latest. Hopefully sooner.

**Comment by Paul-Bob** [comment]
> [@adrianthedev](https://github.com/adrianthedev) Commenting here, following our discussion on Discord. I am looking to implement a filtering on a non existing field (My model has a method Villa.available_between(d1, d2), and a complex query in it). Ideally, I would like the users to select the field, then use the is_within operator and select a range. Is it already possible ? Any workaround ?

@jetienne as curiosity what data types are `d1` and `d2`? Are those dates?

**Comment by jetienne** [comment]
@Paul-Bob Correct, those are dates.

**Comment by Paul-Bob** [comment]
> @Paul-Bob Correct, those are dates.

Date filter already provides an range input when `is_within` condition is selected

**Comment by jetienne** [comment]
@Paul-Bob great news! Could you point to some documentation?

**Comment by Paul-Bob** [comment]
> @Paul-Bob great news! Could you point to some documentation?

https://docs.avohq.io/3.0/dynamic-filters.html#Date
Any `filterable: true` on a date field will automatically generate a dynamic filter with `is_within` condition.

```ruby
field :birthday,
  as: :date,
  first_day_of_week: 1,
  picker_format: "F J Y",
  format: "cccc, d LLLL yyyy", # Wednesday, 10 February 1988
  placeholder: "Feb 24th 1955",
  required: true,
  only_on: [:index],
  filterable: true
```

Test it [here](https://main.avodemo.com/avo/resources/users?filters[birthday][is_within][]=2024-07-23%20to%202024-07-26)

**Comment by jetienne** [comment]
@Paul-Bob This is not a field, this is a complex query, that's why I need to use the new format and be able to scope the query

**Comment by Paul-Bob** [comment]
> @Paul-Bob This is not a field, this is a complex query, that's why I need to use the new format and be able to scope the query

```ruby
dynamic_filter label: "Complex query",
  type: :date,
  query: -> {
    d1, d2 = filter_param.value.split(" to ")
    query.complex_logic(d1, d2)
  }
```

**Comment by Paul-Bob** [comment]
@jetienne any updates on this?

**Comment by Paul-Bob** [comment]
The original request of having a `range` input (`is_within` condition) for number filters was implemented [here.](https://github.com/avo-hq/avo-dynamic_filters/pull/61)

Closing this issue, let's keep track of the conditions improvements [here.](https://github.com/avo-hq/avo/issues/3114)
